python/ql/lib/semmle/python/frameworks/FastApi.qll

```diff
@@ -0,0 +1,353 @@
+  // Response modeling
+  // ---------------------------------------------------------------------------
+  /**
+   * A parameter to a request handler that has a WebSocket type-annotation.
+   */
+  private class WebSocketRequestHandlerParam extends Starlette::WebSocket::InstanceSource,
+    DataFlow::ParameterNode {
+    WebSocketRequestHandlerParam() {
+      this.getParameter().getAnnotation() = Starlette::WebSocket::classRef().getAUse().asExpr() and
+      any(FastApiRouteSetup rs).getARequestHandler().getArgByName(_) = this.getParameter()
+    }
+  }
+
+  /**
+   * Provides models for the `fastapi.Response` class and subclasses.
+   *
+   * See https://fastapi.tiangolo.com/advanced/custom-response/#response.
+   */
+  module Response {
+    /**
+     * Gets the `API::Node` for the manually modeled response classes called `name`.
+     */
+    private API::Node getModeledResponseClass(string name) {
+      name = "Response" and
+      result = API::moduleImport("fastapi").getMember("Response")
+      or
+      // see https://github.com/tiangolo/fastapi/blob/master/fastapi/responses.py
+      name in [
+          "Response", "HTMLResponse", "PlainTextResponse", "JSONResponse", "UJSONResponse",
+          "ORJSONResponse", "RedirectResponse", "StreamingResponse", "FileResponse"
+        ] and
+      result = API::moduleImport("fastapi").getMember("responses").getMember(name)
+    }
+
+    /**
+     * Gets the default MIME type for a FastAPI response class (defined with the
+     * `media_type` class-attribute).
+     *
+     * Also models user-defined classes and tries to take inheritance into account.
+     *
+     * TODO: build easy way to solve problems like this, like we used to have the
+     * `ClassValue.lookup` predicate.
+     */
+    private string getDefaultMimeType(API::Node responseClass) {
+      exists(string name | responseClass = getModeledResponseClass(name) |
+        // no defaults for these.
+        name in ["Response", "RedirectResponse", "StreamingResponse"] and
+        none()
+        or
+        // For `FileResponse` the code will guess what mimetype
+        // to use, or fall back to "text/plain", but claiming that all responses will
+        // have "text/plain" per default is also highly inaccurate, so just going to not
+        // do anything about this.
+        name = "FileResponse" and
+        none()
+        or
+        name = "HTMLResponse" and
+        result = "text/html"
+        or
+        name = "PlainTextResponse" and
+        result = "text/plain"
+        or
+        name in ["JSONResponse", "UJSONResponse", "ORJSONResponse"] and
+        result = "application/json"
+      )
+      or
+      // user-defined subclasses
+      exists(Class cls, API::Node base |
+        cls.getABase() = base.getAUse().asExpr() and
+        // since we _know_ that the base has an API Node, that means there is a subclass
+        // edge leading to the `ClassExpr` for the class.
```

Comment:
Was the two lines below clear to you immediately?

```
cls.getABase() = base.getAUse().asExpr() and
responseClass.getAnImmediateUse().asExpr().(ClassExpr) = cls.getParent()
```

I think the comment tried to explain _why_ that code works... since you're not always able to link up a `Class` with an `API::Node`... but in this case you _are_ able to do that.

Reply:
But I do not see us using the API subclass edge here? We are using the parent edge of a different graph..